### ISSUE TITLE
Add "gopherjs serve" command

### ISSUE THREAD

**Issue by neelance** [gopherjs/gopherjs#121]

**Comment by albrow** [comment]
@neelance, do you want some help with this? I'm very comfortable using the http package in go and built something similar to what you're talking about in [a static blog generator](https://github.com/albrow/scribble) I'm working on. I understand this might be low priority, but it also seems relatively independent and a great candidate for something I could help with. Let me know and I'll work up a PR!

**Comment by neelance** [comment]
Truth is, I'm undecided. I am very happy that you want to contribute. On the other hand I already have quite specific plans for this and it was one of the next things on my todo list, because I need it before starting to promote https://github.com/neelance/dom/tree/master/examples/todomvc . Speaking of promotion, publishing helper libraries, blog posts, etc. would be tremendously useful, because we still need to raise awareness.

**Comment by albrow** [comment]
Totally understand if you think it's easier to just do it yourself. I would love to see gopherjs succeed, so I thought I'd offer to help here and see if it made sense. For now, I'll try and do what I can to help you with promotion. For me that means writing blog posts and building libraries on top of gopherjs.

**Comment by neelance** [comment]
Initial implementation is done. You can try it on the example of https://github.com/neelance/dom . Feedback very welcome. :-)

**Comment by shurcooL** [comment]
Just tried it out. My goal is to deprecate my personal tool and rely on `gopherjs serve`, because it's more general and other people can use it too.

So far it works well. It took me a minute or two to figure out. I've only tried it on my very simple, single `main.go` file GopherJS project.

I like that it serves all of your GOPATH so can use same command to test multiple GopherJS projects (my personal tool only served current directory). But I wonder how that will handle projects that have static assets in working directory of the Go package.

One minor thought/suggestion, I wish it would use port 8080 by default rather than 6060. 6060 is typically used by godoc, and I've always used 8080 for testing personal projects. I know it can be set, which is great, but I just think 8080 is a better default than 6060.

I'll post more feedback as I have it, but so far I think this will work really well!

**Comment by shurcooL** [comment]
It works for Go packages with multiple `.go` files too, so that's great. It looks like one just needs to include `<script src="main.js" type="text/javascript"></script>` in the `index.html` file, that's all (if the default `index.html` doesn't suit you).

**Comment by shurcooL** [comment]
It works for static assets too, but only if I use relative paths.

**Comment by neelance** [comment]
Changed default port to 8080. Yes, assets need relative paths.

**Comment by slimsag** [comment]
Some immediate notes:

1. Navigating through my `$GOPATH` in a browser isn't easy. If it could somehow pick up that I am running `gopherjs serve` from within a specific `$GOPATH` source directory and navigate me there -- all the better.
2. When a compilation error occurs it spits it out to the terminal -- but I expect to be looking constantly at the browser. Maybe the error could be written out to the page being served?

**Comment by shurcooL** [comment]
> \2. When a compilation error occurs it spits it out to the terminal -- but I expect to be looking constantly at the browser. Maybe the error could be written out to the page being served?

Agree.

I recommend the approach my tool has used. Just take the error output of GopherJS and call `console.error` with it. Something like this (except better code style):

```Go
func handleJsError(jsCode string, err error) string {
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return `console.error("` + template.JSEscapeString(err.Error()) + `");`
	}
	return jsCode
}
```

**Comment by slimsag** [comment]
Also I just noticed that the default HTML file has just:

```
<html><head><meta charset="utf-8"><script src="main.js"></script></head></html>
```

There isn't a `<body>` inside of it -- is there any easy way to add one that I am not seeing (I imagine most Gopherjs programs will need a body)?

**Comment by shurcooL** [comment]
Yeah, an empty body should be added to the default `index.html`.

However, your script will not be able to access it, since the script is in head, and body is not yet available.

If you want to do something with the body, you can use the `DOMContentLoaded` listener to delay execution until after body is present:

```Go
func main() {
	document.AddEventListener("DOMContentLoaded", false, func(_ dom.Event) {
		run()
	})
}
```

**Comment by slimsag** [comment]
Is there reason to not move the `<script>` tag into the body itself to simplify the process? E.g.

```
<html>
    <head>
        <meta charset="utf-8">
    </head>
    <body>
        <script src="main.js"></script>
    </body>
</html>
```

Then code like this would work without an event handler:

```
func main() {
	document := js.Global.Get("document")

	...
}
```

**Comment by neelance** [comment]
Yes, the body is left out on purpose, exactly because it is not yet available in the head. The intended way is to create a body element and set it, see `SetBody` function at https://github.com/neelance/dom/blob/846937e2cf9c821adbb01c6f9dc73b1bf48f62a7/dom.go#L253. This has the advantage that the page is not rendered before the script has run, which avoids flickering on page load.

**Comment by neelance** [comment]
@slimsag @shurcooL Is this solution okay for your?

**Comment by slimsag** [comment]
@neelance the [SetBody approach](https://github.com/slimsag/gfx/blob/master/test/webgl_clearing/main.go#L16) works great for me! I wasn't aware that it was that easy (Apologies for the verbosity).

I'd still like to see [compilation errors go to the browser](https://github.com/gopherjs/gopherjs/issues/121#issuecomment-86735912) -- would you like me to file a separate issue for it?

**Comment by shurcooL** [comment]
I agree with and echo what @slimsag said above.

**Comment by neelance** [comment]
@slimsag No need for a separate issue. I'll add this soon.

**Comment by dustywilson** [comment]
My `$GOPATH` has two directories.  My projects are in the second.  `gopherjs serve` only serves from the first.  I don't know how it could be fixed, but could it be fixed?  I really don't want go-gotten packages to be intermingled with my code and I use the two-directory `$GOPATH` to accomplish that.

**Comment by shurcooL** [comment]
@dustywilson I also have `GOPATH` with more than one workspace (for similar reasons, one is for dependencies I don't care about, the other for my own work). This should definitely be supported, since GOPATH is defined as a list of workspaces, not a single workspace path.

I haven't noticed any issues with my own setup and `gopherjs serve`, though. But I know Richard has a single workspace from [past findings](https://github.com/gopherjs/gopherjs/issues/60), so it may not be well tested.

**Comment by shurcooL** [comment]
I've looked at the code, and it's definitely trying to treat GOPATH as a list of workspaces and serve from all of them.

@dustywilson, can you try explicitly going to the URL that you expect to be served? E.g., http://localhost:8080/your/import/path.

I have a feeling that Readdir implementation may not be joining contents of multiple folders, so it may be listing only folders in one GOPATH workspace but not others. But they are still served.

@neelance, are you familiar with https://godoc.org/golang.org/x/tools/godoc/vfs#NameSpace? It may be useful here.

**Comment by dustywilson** [comment]
Quite right, it does seem to point to the right files, even though the index fails to list them.  Thanks!

**Comment by neelance** [comment]
Hey guys,
I recently tried out http://webpack.github.io/ for another project and I was quite impressed. It is easy to use, very fast and the `webpack-dev-server` is really nice for development. So I'm thinking about building a `go-loader` for webpack. Since I'm not a fan of maintaining multiple approaches, I'd like to deprecate `gopherjs serve` if it works out well. What do you think?

**Comment by shurcooL** [comment]
> I'm not a fan of maintaining multiple approaches

That is understandable. I agree that there should be just one official, recommended solution for this in GopherJS README.

My main criteria for what I think would be good is not myself as a user, but other people. I want `gopherjs serve` or equivalent to be so good, that I can use it myself, but primarily it should be good so when I describe how it works to others, all I have to say is "go get gopherjs and then run gopherjs serve, and you're done!"

If I have to explain to others "well, to get this cool functionality that I'm showing you right now, you need to go get gopherjs, but then also install this 3rd party dependency and figure out how to use it", it's not a great solution.

However, if GopherJS itself can itegrate webpack, perhaps that would work? By integrate, I mean perhaps in the README it can describe how to use it, or even embed it so that running `gopherjs serve` actually uses and executes webpack internally.

Having looked at webpack documentation and introduction just now, I think it looks useful and thorough, but at the same time it feels too complex, advanced, and uses too many JavaScript/Node conventions for my liking. Plus it is not written in Go. So my initial reaction is that it doesn't seem like a great `gopherjs serve` replacement (assuming you can make `gopherjs serve` a little better, it's still missing a few nice  features like [propagating build errors to browser console](https://github.com/gopherjs/gopherjs/issues/121#issuecomment-87464537) the way my `gopherjs_serve_http` does).

Given the scope and polish of the webpack project, I was not surprised to see it has 1200~ commits. I am actually working on what is likely effectively a "webpack competitor", but I haven't been able to get it finished enough to release it yet. It would be my solution for #98. But it's written in Go and uses code generation and is based on top of a virtual filesystem paradigm. So I can see the attractiveness of webpack.

**Comment by shurcooL** [comment]
I believe the original issue, described as "Add "gopherjs serve" command", has been completed with the following changes:

- https://github.com/gopherjs/gopherjs/commit/bb7caded937d01202de575f10b09e758a8959330
- https://github.com/gopherjs/gopherjs/commit/e9ec26f835d93166b165d53ef46823e6a11f66e2
- https://github.com/gopherjs/gopherjs/pull/232
- https://github.com/gopherjs/gopherjs.github.io/commit/256347324a580da50a57b074d7eafdcb91737c71

IMO it works great and I'm happy to recommend it to other people rather than my own similar tool (which I can now deprecate).

This issue is becoming longwinded and it's hard to track what actionable things are remaining. I'll close this issue since the original task is complete.

@neelance, in the last few comments you started discussing webpack. Could you make make that a separate issue, if you still want to make that proposal and have a discussion about it?
